runtime/fastly/builtins/fetch-event.cpp

```diff
@@ -0,0 +1,655 @@
+                           "object or a Promise resolving to a Response object as "
+                           "the first argument");
+    JS::RootedObject rejection(cx, PromiseRejectedWithPendingError(cx));
+    if (!rejection)
+      return false;
+    args.rval().setObject(*rejection);
+    return FetchEvent::respondWithError(cx, event);
+  }
+
+  // Step 10.2 (very roughly: the way we handle responses and their bodies is
+  // very different.)
+  JS::RootedObject response_obj(cx, &args[0].toObject());
+
+  // Ensure that all headers are stored client-side, so we retain access to them
+  // after sending the response off.
+  if (Response::is_upstream(response_obj)) {
+    JS::RootedObject headers(cx);
+    headers = RequestOrResponse::headers<Headers::Mode::ProxyToResponse>(cx, response_obj);
+    if (!Headers::delazify(cx, headers))
+      return false;
+  }
+
+  bool streaming = false;
+  if (Response::is_grip_upgrade(response_obj)) {
+    std::string backend(Response::grip_backend(response_obj));
+
+    auto res = host_api::HttpReq::redirect_to_grip_proxy(backend);
+    if (auto *err = res.to_err()) {
+      HANDLE_ERROR(cx, *err);
+      return false;
+    }
+    return true;
+  }
+
+  if (!RequestOrResponse::maybe_stream_body(cx, response_obj, &streaming)) {
+    return false;
+  }
+
+  FetchEvent::set_state(event, streaming ? FetchEvent::State::responseStreaming
+                                         : FetchEvent::State::responseDone);
+  return start_response(cx, response_obj, streaming);
+}
+
+// Steps in this function refer to the spec at
+// https://w3c.github.io/ServiceWorker/#fetch-event-respondwith
+bool response_promise_catch_handler(JSContext *cx, JS::HandleObject event,
+                                    JS::HandleValue promise_val, JS::CallArgs args) {
+  JS::RootedObject promise(cx, &promise_val.toObject());
+
+  fprintf(stderr, "Error while running request handler: ");
+  ENGINE->dump_promise_rejection(args.get(0), promise, stderr);
+
+  // TODO: verify that this is the right behavior.
+  // Steps 9.1-2
+  return FetchEvent::respondWithError(cx, event);
+}
+
+} // namespace
+
+// Steps in this function refer to the spec at
+// https://w3c.github.io/ServiceWorker/#fetch-event-respondwith
+bool FetchEvent::respondWith(JSContext *cx, unsigned argc, JS::Value *vp) {
+  METHOD_HEADER(1)
+
+  // Coercion of argument `r` to a Promise<Response>
+  JS::RootedObject response_promise(cx, JS::CallOriginalPromiseResolve(cx, args.get(0)));
+  if (!response_promise)
+    return false;
+
+  // // Step 2
+  if (!is_dispatching(self)) {
+    JS_ReportErrorUTF8(cx, "FetchEvent#respondWith must be called synchronously from "
+                           "within a FetchEvent handler");
+    return false;
+  }
+
+  // // Step 3
+  if (state(self) != State::unhandled) {
+    JS_ReportErrorUTF8(cx, "FetchEvent#respondWith can't be called twice on the same event");
+    return false;
+  }
+
+  // // Step 4
```

Comment:
There are a few cases where the implementation differs from the original in `runtime/js-compute-runtime/builtins/fetch-event.cpp` in comments only. Could we clean up these comments so that they match a bit closer?